File: src/ltd/rust_lang/utils/list/DefaultList.java
package ltd.rust_lang.utils.list;

public abstract class DefaultList<E> implements MyList<E> {
    protected int size = 0;

    void rangeCheck(int index) {
        if (index >= size || index < 0) outBounds(index);
    }

    void rangeCheckInsert(int index) {
        if (index > size || index < 0) outBounds(index);
    }

    private void outBounds(int index) {
        throw new IndexOutOfBoundsException("index: " + index + ", size: " + size);
    }

    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size() == 0;
    }

    @Override
    public boolean add(E element) {
        insert(size, element);
        return true;
    }
}

File: src/main/kotlin/wiki/mdzz/utils/list/DefaultList.kt
package wiki.mdzz.utils.list

abstract class DefaultList<E> : MyList<E> {
    companion object {
        const val DEFAULT_CAPACITY = 10
        const val ELEMENT_NOT_FOUND = -1
    }

    protected var size: Int = 0

    override fun size(): Int {
        return size
    }

    override val isEmpty: Boolean
        get() = size() == 0

    private fun throwIndexOutOfBoundsException(index: Int): Unit =
        throw IndexOutOfBoundsException("index: $index, size: $size")

    override fun add(element: E?): Boolean {
        add(size(), element)
        return true
    }

    protected open fun rangeCheckForAdd(index: Int) {
        if (index > size() || index < 0) throwIndexOutOfBoundsException(index)
    }

    protected open fun rangeCheck(index: Int) {
        if (index >= size() || index < 0) throwIndexOutOfBoundsException(index)
    }
}

